qcodes/dataset/experiment_settings.py

```diff
@@ -0,0 +1,55 @@
+"""Settings that are indirectly related to experiments."""
+
+from typing import Optional
+
+# The active experiment's exp_id. The default is None.
+active_experiment: Optional[int] = None
+
+
+def _set_active_experiment_id(exp_id: Optional[int]) -> None:
+    """
+    Sets the global active_experiment to the exp_id of a created/ loaded
+    experiment. When a database is loaded, sets to the maximum exp_id in
+    the database. If it is a new database, sets to None.
+
+    Args:
+        exp_id: The exp_id of an experiment.
+    """
+    global active_experiment
+    active_experiment = exp_id
+
+
+def get_active_experiment_id() -> Optional[int]:
+    """
+    Gets the updated global active_experiment.
+
+    Returns:
+        Returns the latest created/ loaded experiment's exp_id in the kernel.
+        If no experiment is started in the kernel and only a database is
+        initialized, the return will be the maximum exp_id in the database
+        or None, if it is a new database.
+    """
+    global active_experiment
+    return active_experiment
+
+
+def reset_active_experiment_id() -> None:
+    """
+    Resets the active_experiment to None.
+    """
+    global active_experiment
+    active_experiment = None
+
+
+def _handle_active_experiment_id_return() -> Optional[int]:
+    """
+    Checks if get_active_experiment_id is an existing exp_id and return it,
+    and if it is None, raise error that no experiment is initialized.
+    """
+    global active_experiment
+    if active_experiment is not None:
+        return active_experiment
+    else:
+        raise ValueError("No experiments found."
+                         " You can create one with:"
+                         " load_or_create_experiment(name, sample_name)")
```

Comment:
My intention was not that we let new_experiment and initialize_database set the active experiment rather I suggested that we implement a function that would handle the logic implemented in dataset. 
Now circular imports may make it tricky to place this function in the experiment module or the settings module where you really would want it so I suggest that we put it in the dataset module for now.

Something like

```python
det get_default_experiment(conn) -> int:
    exp_id = get_active_experiment_id()
    if exp_id is None:
         exp_id = get_last_experiment(conn)
    if exp_id is None:  # if it's still None, then...
         raise ValueError("No experiments found."
                                     "You can start a new one with:"
                                     " new_experiment(name, sample_name)")
     return exp_id
```

Reply:
What I understood from @astafan8 comment was that we should 1. decouple this logic from DataSet, 2. potential use of this logic somewhere else. Your suggestion  somehow addresses point 1, but point 2 might be a problem due to dependency issue, if we want to use this function somewhere else.

Also, there is a small problem if I revert back setting active exp in the database: If you initialize a database having experiments and run `get_active_experiment_id` after that, it returns `None`, which is not correct, despite not being a major issue.

Reply:
While that may be true your solution creates other unwanted couplings. We explicitly do not want to couple the notion of the active_dataset to the sqlite modules. (We have to live with the call to reset in initialize_database for now but that function should be pushed up one level) 

> Also, there is a small problem if I revert back setting active exp in the database: If you initialize a database having experiments and run get_active_experiment_id after that, it returns None, which is not correct, despite not being a major issue.

This is fully intentional. If you want the default experiment you should call get_default_experiment. Active experiment is the one that is explicitly set. If that is not the case you will use the latest experiment

Reply:
OK,
Then I adapt to what you suggested.

Reply:
@jenshnielsen 

Please look at the commit  of "Redisigned the module"
I will modify the rest of files accordingly, but having this new structure seems working fine.
I don't see any good reason why we should make the final function private. What do you think?